macros/src/lib.rs

```diff
@@ -458,3 +461,140 @@ impl Parse for OptionSetter {
+    let mut impl_in = parse_macro_input!(item as ItemImpl);
+
+    // Gather information about each option struct field
+    struct OptInfo {
+        attrs: Vec<Attribute>,
+        type_: Path,
+    }
+    let mut opt_info = HashMap::new();
+    let fields = match &opt_struct.fields {
+        Fields::Named(f) => &f.named,
+        _ => compile_error!(opt_struct.span(), "options struct must have named fields"),
+    };
+    for field in fields {
+        if !matches!(field.vis, Visibility::Public(..)) {
+            continue;
+        }
+        // name
+        let ident = match &field.ident {
+            Some(f) => f.clone(),
+            None => continue,
+        };
+        // doc and cfg attrs
+        let mut attrs = vec![];
+        for attr in &field.attrs {
+            if attr.path().is_ident("doc") || attr.path().is_ident("cfg") {
+                attrs.push(attr.clone());
+            }
+        }
+        // type, unwrapped from `Option`
+        let outer = match &field.ty {
+            Type::Path(ty) => &ty.path,
+            _ => compile_error!(field.span(), "invalid type"),
+        };
+        let type_ = match inner_type(outer, "Option") {
+            Some(Type::Path(ty)) => ty.path.clone(),
+            _ => compile_error!(field.span(), "invalid type"),
+        };
+
+        opt_info.insert(ident, OptInfo { attrs, type_ });
+    }
+
+    // Append utility fns to `impl` block item list
+    let opt_field_type = &opt_struct.ident;
+    impl_in.items.push(parse_quote! {
+        #[allow(unused)]
+        fn options(&mut self) -> &mut #opt_field_type {
+            self.options.get_or_insert_with(<#opt_field_type>::default)
+        }
+    });
+    impl_in.items.push(parse_quote! {
+        /// Set all options.  Note that this will replace all previous values set.
+        pub fn with_options(mut self, value: impl Into<Option<#opt_field_type>>) -> Self {
+            self.options = value.into();
+            self
+        }
+    });
+    // Append setter fns to `impl` block item list
+    for (name, OptInfo { attrs, type_ }) in opt_info {
+        let (accept, value) = if type_.is_ident("String")
+            || type_.is_ident("Bson")
+            || path_eq(&type_, &["bson", "Bson"])
+        {
+            (quote! { impl Into<#type_> }, quote! { value.into() })
+        } else if let Some(t) = inner_type(&type_, "Vec") {
+            (
+                quote! { impl IntoIterator<Item = #t> },
+                quote! { value.into_iter().collect() },
+            )
+        } else {
+            (quote! { #type_ }, quote! { value })
+        };
+        impl_in.items.push(parse_quote! {
+            #(#attrs)*
+            pub fn #name(mut self, value: #accept) -> Self {
+                self.options().#name = Some(#value);
+                self
+            }
+        })
+    }
+
+    // All done.
+    impl_in.to_token_stream().into()
+}
+
+#[import_tokens_attr]
+#[proc_macro_attribute]
+pub fn options_doc(
```

Comment:
This uses the same trick, but in a much simpler way to make the list of setters available in the doc comment for the main action method call itself.  It uses the token stream of the impl block rather than the option struct so we pick up things like `session` and any custom setters we happen to write.